contracts/contracts/chains/optimism/OptimismMessengers.sol

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.12;
+
+import "../../../interfaces/IMessenger.sol";
+
+import "../../RestrictedCalls.sol";
+import "./ICrossDomainMessenger.sol";
+import "./Lib_PredeployAddresses.sol";
+
+abstract contract OptimismMessengerBase is IMessenger, RestrictedCalls {
+    ICrossDomainMessenger public messenger;
+
+    function originalSender() external view returns (address) {
+        return messenger.xDomainMessageSender();
+    }
+
+    function sendMessage(
+        address target,
+        bytes calldata message,
+        uint32 gasLimit
+    ) external restricted(block.chainid, msg.sender) {
+        messenger.sendMessage(target, message, gasLimit);
+    }
+}
+
+contract OptimismL1Messenger is OptimismMessengerBase {
+    constructor(address messenger_) {
+        messenger = ICrossDomainMessenger(messenger_);
+    }
+}
+
+contract OptimismL2Messenger is OptimismMessengerBase {
```

Comment:
Why do we need the distinction between L1 and L2 messenger? I thought you wrote it in a way that it is equal. Is it just to have the predeployed address?

Reply:
For the Optimism case, both are almost identicall, but as we go on adding support for other rollups I suspect we might run into more differences. So I thought it was good to leave some space for that possibility and to clearly indicate which messengers belong to which chain.